lib/fabrik/VirtualHostManager.js

```diff
@@ -0,0 +1,168 @@
+      .tap(() => this.saveVirtualHostInstanceToDeploymentRelation(deploymentName, guid))
+      .catch(err => {
+        logger.error('+-> Failed to create virtual host');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  deleteVirtualHost(deploymentName, guid) {
+    this.verifyFeatureSupport('multi_tenancy');
+    logger.info(`deleting virtual host for deployment '${deploymentName}'...`);
+    return this.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.deprovision(ips, guid))
+      .tap(() => this.deleteVirtualHostInstanceToDeploymentRelation(deploymentName, guid))
+      .catch(err => {
+        logger.error('+-> Failed to delete virtual host');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  createBinding(deploymentName, guid, binding) {
+    this.verifyFeatureSupport('credentials');
+    logger.info(`Creating binding '${binding.id}' for deployment '${deploymentName}', virtual host '${guid}'...`);
+    logger.info('+-> Binding parameters:', binding.parameters);
+    return this.getDeploymentIps(deploymentName)
+      .then(ips => this.agent.createCredentials(ips, guid, binding.parameters))
+      .tap(credentials => this.createBindingProperty(deploymentName, binding.id, _.set(binding, 'credentials', credentials)))
+      .tap(() => {
+        const bindCreds = _.cloneDeep(binding.credentials);
+        utils.maskSensitiveInfo(bindCreds);
+        logger.info(`+-> Created binding:${JSON.stringify(bindCreds)}`);
+      })
+      .catch(err => {
+        logger.error('+-> Failed to create binding');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  deleteBinding(deploymentName, guid, id) {
+    this.verifyFeatureSupport('credentials');
+    logger.info(`Deleting binding '${id}' for deployment '${deploymentName}' , virtual host '${guid}'...`);
+    return Promise
+      .all([
+        this.getDeploymentIps(deploymentName),
+        this.getBindingProperty(deploymentName, id)
+      ])
+      .spread((ips, binding) => this.agent.deleteCredentials(ips, guid, binding.credentials))
+      .then(() => this.deleteBindingProperty(deploymentName, id))
+      .tap(() => logger.info('+-> Deleted service binding'))
+      .catch(err => {
+        logger.error('+-> Failed to delete binding');
+        logger.error(err);
+        throw err;
+      });
+  }
+
+  saveVirtualHostInstanceToDeploymentRelation(deploymentName, id) {
+    return this.mapper
+      .createVirtualHostRelation(deploymentName, `virtual_host_instance-${id}`)
+      .catchThrow(BadRequest, new BadRequest(`Virtual host instance with ID ${id} already exists on deployment ${deploymentName}`));
+  }
+
+  deleteVirtualHostInstanceToDeploymentRelation(deploymentName, id) {
+    return this.mapper
+      .deleteVirtualHostRelation(`virtual_host_instance-${id}`)
+      .catchThrow(BadRequest, new BadRequest(`Virtual host instance with ID ${id} does not exists on deployment ${deploymentName}`));
```

Comment:
Delete operations must be idempotent. If a delete request for an already deleted request comes in, must return back with HTTP STATUS 410 - GONE & should not throw an error.

Reply:
From here you could rather throw `ServiceBindingNotFound` and in the controller you can treat this exception with HTTP Status 410 - look [here for ref](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/master/lib/controllers/ServiceBrokerApiController.js#L211)